File: solutions/_95_unique_binary_search_trees_ii/solution.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        self.memo = {}
        return self.generate(1, n) if n else []

    def generate(self, start, end):
        if start > end:
            return [None]
        if (start, end) in self.memo:
            return self.memo[(start, end)]
        res = []
        for val in range(start, end + 1):
            for left_tree in self.generate(start, val - 1):
                for right_tree in self.generate(val + 1, end):
                    node = TreeNode(val)
                    node.left = left_tree
                    node.right = right_tree
                    res.append(node)

        self.memo[(start, end)] = res
        return res


# https://leetcode.com/problems/unique-binary-search-trees-ii/

File: solutions/_95_unique_binary_search_trees_ii/Solution.java

package _95_unique_binary_search_trees_ii;

import java.util.*;

class Solution {
    HashMap<String, List<TreeNode>> memo = new HashMap<>();

    public List<TreeNode> generateTrees(int n) {
        if (n == 0) {
            return new ArrayList<>();
        }
        return generate(1, n);
    }

    private List<TreeNode> generate(int start, int end) {
        List<TreeNode> result = new ArrayList<>();
        if (start > end) {
            result.add(null);
            return result;
        } else if (memo.containsKey(start + "-" + end)) {
            return memo.get(start + "-" + end);
        }

        for (int val = start; val <= end; val++) {
            for (TreeNode leftTree : generate(start, val - 1)) {
                for (TreeNode rightTree : generate(val + 1, end)) {
                    TreeNode node = new TreeNode(val);
                    node.left = leftTree;
                    node.right = rightTree;
                    result.add(node);
                }
            }
        }
        memo.put(start + "-" + end, result);
        return result;
    }
}

/*
 * https://leetcode.com/problems/unique-binary-search-trees-ii/
 */